### ISSUE TITLE
chap06_Keras_Model_keras_mnist_DNN_28_28

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#46]
"""
Tensorflow2.0 keras + Mnist(0~9) + Flatten layer

1차 방법 : 1차원 : (28x28) -> 784
2차 방법 : 2차원 : 28 x 28 -> flatten 적용
"""

import tensorflow as tf # ver2.x
from tensorflow.keras.datasets.mnist import load_data # ver2.x dataset
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import minmax_scale # x변수 전처리
from tensorflow.keras.utils import to_categorical # 원핫코딩과 비슷한것, y변수 전처리
from tensorflow.keras import Sequential # model 생성
from tensorflow.keras.layers import Dense, Flatten # layer 생성
from tensorflow.keras.models import load_model # model save/load
from sklearn.metrics import accuracy_score

### 1. x, y 공급 data 
(x_train, y_train), (x_val, y_val) = load_data()
x_train.shape # image # (60000, 28, 28) # 3없음:색깔없음
y_train.shape # labels # (60000,)

- x변수 전처리 : 정규화
x_train[0] # 0~255숫자가 28x28행렬형태
x_train = x_train/255.  # 255. 으로 나누어줌으로써 실수로 변경
x_val = x_val/255.

'''
- 2d -> 1d
x_train = x_train.reshape(-1, 784)
x_val = x_val.reshape(-1, 784)
'''

- y변수 : one hot encoding
y_train = to_categorical(y_train) # y변수 전처리 : one hot encoding
y_val = to_categorical(y_val)
y_train.shape # (60000, 10)

### 2. keras model 생성
model = Sequential()
model # object info

### 3. model layer
'''
model.add(Dense(node수, input_shape, activation)) # hidden layer1
model.add(Dense(node수, activation)) # hidden layer1 ~ n
'''
input_shape = (28, 28)

- Flatten layer  2d(28,28) -> 1d(784)
model.add(Flatten(input_shape = input_shape)) # 0층

- hidden layer1 = [784,128]
model.add(Dense(128, activation = 'relu')) # 1층
- hidden layer2 = [128,64]
model.add(Dense(64, activation = 'relu')) # 2층
- hidden layer3 = [64,32]
model.add(Dense(32, activation = 'relu')) # 2층
- output layer2 = [32,10]
model.add(Dense(10, activation = 'softmax')) # 3층

### 4. model compile : 학습환경 설정
model.compile(optimizer = 'adam', # 최적화 알고리즘(lr 생략)
              loss = 'categorical_crossentropy', # 손실
              metrics=['accuracy']) # 평가 방법

- layer 확인
model.summary()

### 5. model training : train(112) vs val(38)
model.fit(x=x_train, y=y_train,  # 학습용
          epochs=10,
          verbose=1, # 학습과정 횟수
          validation_data=(x_val, y_val) # 평가용
          )

### 6. model evaluation : 모델 검증
model.evaluate(x=x_val, y=y_val)
[0.08544410765171051, 0.9790999889373779]

### 7. model save/load
model.save("keras_model_mnist2.h5")
print("model save")
특별하게 경로지정하지 않으면 현재경로에 파일로 저장됨.

new_model = load_model("keras_model_mnist2.h5")
print(new_model) #object info

### 8. model test : new dataset
y_pred = new_model.predict(x_val) # 예측치 : 확률값
y_true = y_val # 관측치 : one hot
비교할 수 없으므로 이것을 맞추어야 한다.

y_pred = tf.argmax(y_pred, axis=1)
y_true = tf.argmax(y_true, axis=1)

from sklearn.metrics import accuracy_score

acc = accuracy_score(y_true, y_pred)
print("accuracy =", acc) # accuracy = 0.9791
